designs/move-cluster-specific-code-out-of-manager.md

```diff
@@ -0,0 +1,242 @@
+	// sensitive and shouldn't be exposed publicly.
+	// If the simple path -> handler mapping offered here is not enough, a new http server/listener should be added as
+	// Runnable to the manager via Add method.
+	AddMetricsExtraHandler(path string, handler http.Handler) error
+
+	// AddHealthzCheck allows you to add Healthz checker
+	AddHealthzCheck(name string, check healthz.Checker) error
+
+	// AddReadyzCheck allows you to add Readyz checker
+	AddReadyzCheck(name string, check healthz.Checker) error
+
+	// Start starts all registered Controllers and blocks until the Stop channel is closed.
+	// Returns an error if there is an error starting any controller.
+	// If LeaderElection is used, the binary must be exited immediately after this returns,
+	// otherwise components that need leader election might continue to run after the leader
+	// lock was lost.
+	Start(<-chan struct{}) error
+
+	// GetWebhookServer returns a webhook.Server
+	GetWebhookServer() *webhook.Server
+}
+```
+
+Additionally, we will export a new `MultiClusterManager` that is intended to be used for
+multi-cluster controllers:
+
+```go
+type MultiClusterManager interface {
+	Manager
+	// Add another named cluster to the MultiClusterManager. The name
+	// must be unique. The MultiClusterManager will wait for all clusters
+	// caches to be started before starting anytything else.
+	AddCluster(config *rest.Config, name string) error
+	// GetClusters returns all Clusters this MultiClusterManager knows about.
+	// The cluster used to construct the MultiClusterManager is named `primary`
+	// and will be used for LeaderElection, if enabled.
+	GetClusters() map[string]clusterconnector.ClusterConnector
+}
+```
+
+Furthermore, during startup, the `Manager` will use type assertion to find `ClusterConnector`s
+to be able to start their caches before anything else:
+
+```go
+if cc, isClusterConnector:= runnable.(clusterconnector.ClusterConnector); isClusterConnector {
+	m.caches = append(m.caches, cc.GetCache())
+}
+```
+
+```go
+for idx := range cm.caches {
+	go func(idx int) {cm.caches[idx].Start(cm.internalStop)}
+}
+
+for _, cache := range cm.caches {
+	cache.WaitForCacheSync(cm.internalStop)
+}
+
+// Start all other runnables
+```
+
+## Example
+
+Below is a sample `reconciler` that will create a secret in a `mirrorCluster` for each
+secret found in `referenceCluster` if none of that name already exists. To keep the sample
+short, it won't compare the contents of the secrets.
+
+```go
+type secretMirrorReconciler struct {
+	referenceClusterClient, mirrorClusterClient client.Client
+}
+
+func (r *secretMirrorReconciler) Reconcile(r reconcile.Request)(reconcile.Result, error){
+	s := &corev1.Secret{}
+	if err := r.referenceClusterClient.Get(context.TODO(), r.NamespacedName, s); err != nil {
+		if kerrors.IsNotFound{ return reconcile.Result{}, nil }
+		return reconcile.Result, err
+	}
+
+	if err := r.mirrorClusterClient.Get(context.TODO(), r.NamespacedName, &corev1.Secret); err != nil {
```

Comment:
Tangent: Not that we need to tackle it here, but we could even create a multicluster client that (ab)used the unused `ClusterName` field (maybe dangerous), or used a client option, context field, etc to avoid needing multiple clients.

Reply:
That sounds like a great idea (but I would prefer to keep that as an orthogonal work item)